contrib: skip missing binaries in gen-manpages

andremralves opened the pull request:
Instead of stopping the execution of gen-manpages.py when a binary is not found, continue generating manpages for the available binaries and skip the missing ones.

A new argument, `--skip-missing-binaries`, has been added to enable this behavior.

```sh
➜  bitcoin git:(fix-gen-manpages) ✗ ./contrib/devtools/gen-manpages.py --help
usage: gen-manpages.py [-h] [-s]

options:
  -h, --help            show this help message and exit
  -s, --skip-missing-binaries
                        skip generation for binaries that are not found

```

closes #30985

This PR also includes an error prompt if no binaries are found in the build path.

willcl-ark commented:
Concept ACK.

I wonder if it's not easier to minimize what is being caught in the IOError try block and just remove missing bins from the `BINARIES` list? This list is used later on in the script too so in my opinion that is probably the better option. Otherwise we will end up referencing non-existing binaries/manpages in the existing manpages. Something like:

```diff
diff --git a/contrib/devtools/gen-manpages.py b/contrib/devtools/gen-manpages.py
index 92acd9a4037..0e0d31807bd 100755
--- a/contrib/devtools/gen-manpages.py
+++ b/contrib/devtools/gen-manpages.py
@@ -39,7 +39,8 @@ for relpath in BINARIES:
         r = subprocess.run([abspath, "--version"], stdout=subprocess.PIPE, check=True, text=True)
     except IOError:
         print(f'{abspath} not found or not an executable', file=sys.stderr)
-        sys.exit(1)
+        BINARIES.remove(relpath)
+        continue
     # take first line (which must contain version)
     verstr = r.stdout.splitlines()[0]
     # last word of line is the actual version e.g. v22.99.0-5c6b3d5b3508
```

I also notice running this post-cmake *without specifying `BUILDDIR`* results in a slightly incorrect error due to the new binary locations:

```
₿ ./contrib/devtools/gen-manpages.py
/home/will/src/bitcoin/src/bitcoind not found or not an executable. Skipping...
/home/will/src/bitcoin/src/bitcoin-cli not found or not an executable. Skipping...
/home/will/src/bitcoin/src/bitcoin-tx not found or not an executable. Skipping...
/home/will/src/bitcoin/src/bitcoin-wallet not found or not an executable. Skipping...
/home/will/src/bitcoin/src/bitcoin-util not found or not an executable. Skipping...
/home/will/src/bitcoin/src/qt/bitcoin-qt not found or not an executable. Skipping...
Traceback (most recent call last):
  File "/home/will/src/bitcoin/./contrib/devtools/gen-manpages.py", line 63, in <module>
    footer.write('\n'.join(versions[0][2]).strip())
                           ~~~~~~~~^^^
IndexError: list index out of range
```

Whilst you are in here, perhaps we could return early if `versions` is empty, optionally prompting folks to set `BUILDDIR`?

We could change the default paths, but as people may use different build dirs I'm not sure if this is worth it.

laanwj commented:
i'm divided on this. There's a use-case, for sure, but this script is mainly used to generate the manpages for a release. And we've had exactly the opposite issue where conditional behavior resulted in incomplete manual pages.
(Prompting me to open #17506)
So not sure it should be the default. As an option, sure.

andremralves commented:
The main issue this PR addresses is that currently, no man pages are generated if a single binary is missing. The goal is to skip the missing binary and generate man pages for the others, as was done in the previous shell script.

If I understand correctly, you want to ensure no information is lost in the man pages, right? In that case, I can simply remove the following line:
```py
42          BINARIES.remove(relpath)
```
This way, the existing binaries will have their man pages generated, and all the other man pages will still be referenced at the bottom, even if the binaries are missing.

laanwj commented:
Kind of. The problem is that the command will still finish successfully in case of an incomplete build (or build configuration), while the before a release it's critical to update all the pages.

This is why i suggested to add this as an option eg `--skip-missing-binaries`.

> This way, the existing binaries will have their man pages generated, and all the other man pages will still be referenced at the bottom, even if the binaries are missing.

That is an improvement, but doesn't entirely address the concern.

andremralves commented:
I probably don’t have enough understanding of the release process to know exactly how this change could affect it, but I assume that having gen-manpages.py exit with an error is important somewhere during the process.

I implemented your suggestion by adding the --skip-missing-binaries option to the command.

Thanks for the feedback! What do you think of it now?

fanquake commented:
I wonder if a better approach here would be to further integrate the manpage generation into the build system; given that the two issues of "which binaries exist" and "where do they exist" are already solved in that case. This should also remove the dep on `git`.

BrandonOdiwuor reviewed (commented):
Concept ACK
Adding `--skip-missing-binaries` as an optional flag seems reasonable.

rkrux commented on contrib/devtools/gen-manpages.py:
```diff
@@ -68,6 +68,10 @@
     versions.append((abspath, verstr, copyright))
 
+if not versions:
+    print(f'No binaries found in {builddir}. Set BUILDDIR to specify the correct build path.')
```
```
Set BUILDDIR to specify the correct build path.
```

Is the intent to make the user explicitly set BUILDDIR? Not aware of the usual use cases but as a user, my preference is to just run the script to generate the man pages w/o needing to set an env var. I'd be ok with this error as well.

```
print(f'No binaries found in {builddir}. Please specify the correct build path.')
```

andremralves replied:
The most common use case is running the script without setting BUILDDIR, in which case the default builddir, `/build`, is used. However, if no binaries are found in `/build` we remind the user to set BUILDDIR to specify the correct path.

rkrux replied:
> if no binaries are found in /build

This could also be because the user might not have built the binaries yet even in the case of default build path.

A more verbose message could be:
```
'No binaries found in {builddir}. Please ensure the binaries are present in {builddir}, or set another build path using the BUILDDIR env variable.' 
```

andremralves replied:
> This could also be because the user might not have built the binaries yet even in the case of default build path.

Yes, I haven't added it to the message because I though it would be too verbose, but I can reconsider it.

andremralves replied:
I've updated it, thank for the suggestion.

rkrux reviewed (approved):
Coming here from https://github.com/bitcoin/bitcoin/pull/31255, putting my two cents in.

I had to generate man pages while reviewing a PR and found that the script threw error even in the default case due to the new build process that moved the binaries in `/build`, so a strong ACK for 0194a638d31fef7e8193c50843c11d2c2ee212a3 because IMO the script should work in the default case w/o needing to set the `BUILDDDIR`.

As I read from @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/30986#issuecomment-2435085619) that the main intent of this script is to generate man pages for a release, I'd assume all the binaries would be found because they are part of the release? Having said this, the new `--skip-missing-binaries` arg could be useful in the cases when not everyone will have all the binaries. Eg: I didn't use to have the GUI earlier.

All in all ACK for f0806188dcde531a401bdfeaaed8c04801f4b5f0

andremralves commented on contrib/devtools/gen-manpages.py:
```diff
@@ -16,6 +17,18 @@
+    "--skip-missing-binaries",
+    action="store_true",
+    default=False,
+    help="skip generation for binaries that are not found",
```
Good suggestion. Thanks!

andremralves committed:
gen-manpages: implement --skip-missing-binaries

With --skip-missing-binaries, instead of stopping the execution of
gen-manpages.py when a binary is not found, continue generating
manpages for the available binaries and skip the missing ones.

laanwj reviewed (approved):
Code review ACK 4bbd28baf33382231f4f1dab20681c05f9915af2

rkrux reviewed (approved):
re-ACK 4bbd28baf33382231f4f1dab20681c05f9915af2

Minimal range diff.

<details>
<summary>git range-diff f080618...4bbd28b</summary>

```
1:  7060d64bf7 ! 1:  299e2220e9 gen-manpages: implement --skip-missing-binaries
    @@ contrib/devtools/gen-manpages.py: BINARIES = [
     +    "--skip-missing-binaries",
     +    action="store_true",
     +    default=False,
    -+    help="skip generation for binaries that are not found",
    ++    help="skip generation for binaries that are not found in the build path",
     +)
     +args = parser.parse_args()
     +
2:  0194a638d3 = 2:  c32104f544 gen-manpages: Update default build path
3:  f0806188dc ! 3:  4bbd28baf3 gen-manpages: Prompt error if no binaries are found
    @@ contrib/devtools/gen-manpages.py: for relpath in BINARIES:
          versions.append((abspath, verstr, copyright))

     +if not versions:
    -+    print(f'No binaries found in {builddir}. Set BUILDDIR to specify the correct build path.')
    ++    print(f'No binaries found in {builddir}. Please ensure the binaries are present in {builddir}, or set another build path using the BUILDDIR env variable.')
     +    sys.exit(1)
     +
      if any(verstr.endswith('-dirty') for (_, verstr, _) in versions):
```
</details>

willcl-ark reviewed (approved):
ACK 4bbd28baf33382231f4f1dab20681c05f9915af2

Much better to add this behind a default-false flag

achow101 commented on contrib/devtools/gen-manpages.py:
```diff
@@ -40,7 +40,7 @@
 
 # Get input and output directories.
-builddir = os.getenv('BUILDDIR', topdir)
+builddir = os.getenv('BUILDDIR', os.path.join(topdir, 'build'))
```
In c32104f5449bbea89a3fde5b99c97f7de6914829 "gen-manpages: Update default build path"

The documentation states that this script can work from any directory for in-tree builds, but this change would break that (although we do not currently do in-tree builds). It further states that `BUILDDIR` should be set for out-of-tree builds. Either the documentation should be updated, or this default should not be changed. I prefer the latter.

andremralves replied:
Hmm, okay. I think it's better to remove it from this PR and maybe address the subject in another issue so it doesn’t interfere with the other commits.

So, the PR will include only:
- The --skip-missing-binaries option
- The error prompt to inform the user if no binaries are found in the build directory path.

laanwj replied:
Yes, it's unrelated to the scope of the PR anyway.

andremralves committed:
gen-manpages: Prompt error if no binaries are found

laanwj reviewed (approved):
re-ACK ee6185372fc317d3948690997117e42f6b79a5ff
